clang/test/Driver/flang/classic_flang.f95

```diff
@@ -1,28 +1,44 @@
-! Check that the driver can invoke flang1 and flang2 to compile Fortran with
-! --driver-mode=flang (default when the file extension is .f95).
-
 ! REQUIRES: classic_flang
 
-! RUN: %clang -target x86_64-unknown-linux-gnu -integrated-as -c %s -### 2>&1 \
+! Check that the driver invokes flang1 correctly for preprocessed free-form
+! Fortran code. Also check that the backend is invoked correctly.
+
+! RUN: %clang --driver-mode=flang -target x86_64-unknown-linux-gnu -c %s -### 2>&1 \
 ! RUN:   | FileCheck --check-prefix=CHECK-OBJECT %s
 ! CHECK-OBJECT: flang1
-! CHECK-OBJECT: flang2
+! CHECK-OBJECT-NOT: "-preprocess"
+! CHECK-OBJECT-SAME: "-freeform"
+! CHECK-OBJECT-NEXT: flang2
 ! CHECK-OBJECT-SAME: "-asm" [[LLFILE:.*.ll]]
-! CHECK-OBJECT-NOT: cc1as
-! CHECK-OBJECT: clang
-! CHECK-OBJECT-SAME: -cc1
+! CHECK-OBJECT-NEXT: {{clang.* "-cc1"}}
 ! CHECK-OBJECT-SAME: "-o" "classic_flang.o"
 ! CHECK-OBJECT-SAME: "-x" "ir"
 ! CHECK-OBJECT-SAME: [[LLFILE]]
 
-! RUN: %clang -target x86_64-unknown-linux-gnu -integrated-as -S %s -### 2>&1 \
+! Check that the driver invokes flang1 correctly when preprocessing is
+! explicitly requested.
+
+! RUN: %clang --driver-mode=flang -target x86_64-unknown-linux-gnu -E %s -### 2>&1 \
+! RUN:   | FileCheck --check-prefix=CHECK-PREPROCESS %s
+! CHECK-PREPROCESS: flang1
+! CHECK-PREPROCESS-SAME: "-preprocess"
+! CHECK-PREPROCESS-SAME: "-es"
+! CHECK-PREPROCESS-SAME: "-pp"
+! CHECK-PREPROCESS-NOT: flang1
+! CHECK-PREPROCESS-NOT: flang2
+! CHECK-PREPROCESS-NOT: clang
```

Comment:
I recommend - where you can - enclosing all of these (and elsewhere) in quotes. We've seen an issue in the past where tests fail because the working directory contains, e.g. `clang`, because maybe the directory name is based on a branch name, or a user name, and suddenly the test fails.

Reply:
OK I will turn these into regular expressions.

Reply:
Fixed.